src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentAddressToEdifactMapper.java

```diff
@@ -0,0 +1,132 @@
+
+@Component
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class AmendmentAddressToEdifactMapper extends AmendmentToEdifactMapper {
+
+    @Override
+    Segment mapPatches(AmendmentBody amendmentBody) {
+        var patches = amendmentBody.getJsonPatches();
+
+        var houseName = patches.getHouseName()
+            .map(AmendmentPatch::getFormattedSimpleValue)
+            .orElse(null);
+        var numberOrRoadName = patches.getNumberOrRoadName()
+            .map(AmendmentPatch::getFormattedSimpleValue)
+            .orElse(null);
+        var locality = patches.getLocality()
+            .map(AmendmentPatch::getFormattedSimpleValue)
+            .orElse(null);
+        var postalTown = patches.getPostTown()
+            .map(AmendmentPatch::getFormattedSimpleValue)
+            .orElse(null);
+        var county = patches.getCounty()
+            .map(AmendmentPatch::getFormattedSimpleValue)
+            .orElse(null);
+        var postalCode = patches.getPostalCode()
+            .map(AmendmentPatch::getFormattedSimpleValue)
+            .orElse(null);
+
+        return PersonAddress.builder()
+            .addressLine1(houseName)
+            .addressLine2(numberOrRoadName)
+            .addressLine3(locality)
+            .addressLine4(postalTown)
+            .addressLine5(county)
+            .postalCode(postalCode)
+            .build();
+    }
+
+    @Override
+    boolean shouldCreateSegment(AmendmentBody amendmentBody) {
+        var patches = amendmentBody.getJsonPatches();
+        return Stream.of(
+            patches.getHouseName(),
+            patches.getNumberOrRoadName(),
+            patches.getLocality(),
+            patches.getPostTown(),
+            patches.getCounty(),
+            patches.getPostalCode())
+            .anyMatch(Optional::isPresent);
+    }
+
+    @Override
+    protected void validatePatches(JsonPatches patches) {
+        validatePostalCodePatchIfExists(patches);
+        checkIfThereAreAllFiveAddressLinesPatches(patches);
+        checkNoPostTownPatchForRemoveOperation(patches);
+        checkLocalityPostTownAndCountyAllEmptyBlankOrAllNotEmptyBlank(patches);
+    }
+
+    private void validatePostalCodePatchIfExists(JsonPatches patches) {
+        if (patches.getPostalCode().isPresent()) {
+            validateNonEmptyValues(Collections.singletonList(patches.getPostalCode()));
+        }
+    }
+
+    private void checkIfThereAreAllFiveAddressLinesPatches(JsonPatches patches) {
+        if(!Stream.of(
+            patches.getHouseName(),
+            patches.getNumberOrRoadName(),
+            patches.getLocality(),
+            patches.getPostTown(),
+            patches.getCounty())
+            .allMatch(Optional::isPresent)) {
+            throw new FhirValidationException("All five address lines must be provided for amendment");
+        }
+    }
+
+    private void checkLocalityPostTownAndCountyAllEmptyBlankOrAllNotEmptyBlank(JsonPatches patches) {
```

Comment:
This method is really complicated. Would that work:
````java
private void checkLocalityPostTownAndCountyAllEmptyBlankOrAllNotEmptyBlank(JsonPatches patches) {
    if(hasLocalityOrCountyRemovalOperation(patches)) {
        return;
    }

    var localityString = patches.getLocality().get().getValue().get();
    var countyString = patches.getCounty().get().getValue().get();
    var postTownString = patches.getPostTown().get().getValue().get();

    if(allStringsAreBlank(localityString, countyString, postTownString)) {
        return;
    }

    if (anyIsBlank(localityString, countyString, postTownString)) {
        throw new FhirValidationException("If at least one of the Address - Locality, Address - Post Town and Address County " +
            "fields is amended for a patient, then the values held for all three of these fields MUST be provided. Actual state: " +
            "Locality: " + localityString + ", Post Town: " + postTownString + ", County: " + countyString);
    }
}

private boolean allStringsAreBlank(String localityString, String countyString, String postTownString) {
    return localityString.isBlank() && countyString.isBlank() && postTownString.isBlank();
}

private boolean anyIsBlank(String localityString, String countyString, String postTownString) {
    return localityString.isBlank() || countyString.isBlank() || postTownString.isBlank();
}
````

Reply:
I will check your solution if it meets all requirements

Reply:
No, it is logically incorrect, it does nto check the consistency of this three values. Moreover we should not check here them being blank, but not null.